keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+An important note for this solution is the requirement that both sharable and
+instance-specific configuration has the same Kind and APIVersion. This avoids
+the complexities of trying to merge across Kinds, or perform conversions before
+merging.
+
+### Code merge example
+
+In the example above, it's possible to control the patch strategy used in the
+configuration struct field tagging:
+
+```golang
+// External Kubelet configuration API
+// k8s.io/kubelet/config/v1beta1/types.go
+
+type KubeletConfiguration struct {
+    ClusterDNS []string `json:"clusterDNS" strategyMerge:"replace"`
+    FeatureGates map[string]bool `json:"featureGates,omitempty" strategyMerge:"merge"`
+}
+```
+
+The StrategicMergePatch operates directly on the YAML but uses the strategies
+specified in struct tags on the passed-in object. An empty configuration object
+can be passed to the patch function to specify the desired patch strategies.
+
+```golang
+// Some general configuration utilities
+// utils/patch.go
+...
+import (
+    "sigs.k8s.io/yaml"
+)
+...
+
+// ConvertYAMLToJSON convert the YAML configuration to JSON.
+func ConvertYAMLToJSON(yamlSharable, yamlInstance []byte) ([]byte, []byte, error) {
+    jsonSharable, err := yaml.YAMLToJSON(yamlSharable)
+    If err != nil {
+        return nil, nil, err
+    }
+
+jsonInstance, err := yaml.YAMLToJSON(yamlInstance)
+    If err != nil {
+        return nil, nil, err
+    }
+
+    return jsonSharable, jsonInstance, nil
+}
+
+// PatchConfiguration applies a strategic merge patch and returns the patched JSON result,
+// it is necessary to convert both configurations from YAML to JSON before applying the merge.
+func PatchConfiguration(jsonSharable, jsonInstance []byte) []byte, error {
+    obj := &kubeletv1beta1.KubeletConfiguration{}
+
+    patched, err := strategicpatch.StrategicMergePatch(jsonSharable, jsonInstance, obj)
+    if err != nil {
+        return []byte{}, err
+    }
+
+    return patched, err
+}
+```
+
+The steps for this merges is simpler than the alternative approach:
+
+1. Convert shared.yaml configuration from YAML to JSON.
+2. Convert instance.yaml configuration from YAML to JSON.
+3. Patch the shared JSON configuration with the instance JSON, the
+   apimachinery strategic merge patch.
+
+In the diagram below it's possible to have a better visualization of the
+strategic merge flow:
+
+![component-smp-merge](component-smp-merge.png)
+
+### Limitations
+
+This approach has the following limitations. We may decide to address these in the
+future but it is not a goal of this KEP.
+
+* Users can't currently choose the merge strategy. It might make sense for users
+  to be able to choose, since they are the authors of the configuration.
+* This merge strategy requires both objects to have the same Kind. In the future
+  it may turn out to be a good idea to more carefully categorize fields as shared
+  or instance-specific.
+* This merge strategy requires both objects to have the same API version. In the
```

Comment:
Yes, we can do this with a dedicated `--convert-config=(optional desired API version)` flag or command on the component, reading stdin, and writing to stdout.

Reply:
Maybe in the future, for now conversions do defaulting and I'd worry that could interfere with the meaning of the patch. I also don't think it's worth the complexity, it's likely the same person writing both the shared config and the late-bound patch, they can coordinate the versions.